File: client/typescript/src/commands.ts
/** Command builders — produce step JSON for the REST API. */

import { ElementSelector } from "./models";

type Step = Record<string, unknown>;
type SelectorValue = string | Record<string, unknown>;

function selectorValue(opts: {
  text?: string;
  id?: string;
  index?: number;
  selector?: ElementSelector;
  enabled?: boolean;
  checked?: boolean;
  focused?: boolean;
  selected?: boolean;
}): SelectorValue {
  const d: Record<string, unknown> = {};
  if (opts.selector) Object.assign(d, opts.selector.toDict());
  if (opts.text != null) d.text = opts.text;
  if (opts.id != null) d.id = opts.id;
  if (opts.index != null) d.index = String(opts.index);
  if (opts.enabled != null) d.enabled = opts.enabled;
  if (opts.checked != null) d.checked = opts.checked;
  if (opts.focused != null) d.focused = opts.focused;
  if (opts.selected != null) d.selected = opts.selected;
  // Compact form: text-only selector → plain string
  const keys = Object.keys(d);
  if (keys.length === 1 && keys[0] === "text") return d.text as string;
  return d;
}

export function tapOn(opts: {
  text?: string;
  id?: string;
  index?: number;
  selector?: ElementSelector;
  longPress?: boolean;
  waitUntilVisible?: boolean;
  retryIfNoChange?: boolean;
  enabled?: boolean;
  checked?: boolean;
  focused?: boolean;
  selected?: boolean;
  optional?: boolean;
  timeout?: number;
  label?: string;
}): Step {
  const step: Step = { type: "tapOn" };
  step.selector = selectorValue(opts);
  if (opts.longPress) step.longPress = true;
  if (opts.waitUntilVisible != null) step.waitUntilVisible = opts.waitUntilVisible;
  if (opts.retryIfNoChange != null) step.retryTapIfNoChange = opts.retryIfNoChange;
  if (opts.optional) step.optional = true;
  if (opts.timeout != null) step.timeout = opts.timeout;
  if (opts.label != null) step.label = opts.label;
  return step;
}

export function inputText(text: string, label?: string): Step {
  const step: Step = { type: "inputText", text };
  if (label != null) step.label = label;
  return step;
}

export function eraseText(characters?: number, label?: string): Step {
  const step: Step = { type: "eraseText" };
  if (characters != null) step.charactersToErase = characters;
  if (label != null) step.label = label;
  return step;
}

export function pressKey(code: string, label?: string): Step {
  const step: Step = { type: "pressKey", key: code };
  if (label != null) step.label = label;
  return step;
}

export function back(label?: string): Step {
  const step: Step = { type: "back" };
  if (label != null) step.label = label;
  return step;
}

export function scroll(label?: string): Step {
  const step: Step = { type: "scroll" };
  if (label != null) step.label = label;
  return step;
}

export function swipe(direction: string, durationMs: number = 400, label?: string): Step {
  const step: Step = { type: "swipe", direction: direction.toUpperCase(), duration: durationMs };
  if (label != null) step.label = label;
  return step;
}

export function hideKeyboard(strategy?: string, label?: string): Step {
  const step: Step = { type: "hideKeyboard" };
  if (strategy != null) step.strategy = strategy;
  if (label != null) step.label = label;
  return step;
}

export function waitForAnimationToEnd(label?: string): Step {
  const step: Step = { type: "waitForAnimationToEnd" };
  if (label != null) step.label = label;
  return step;
}

export function assertVisible(opts: {
  text?: string;
  id?: string;
  selector?: ElementSelector;
  timeoutMs?: number;
  optional?: boolean;
  label?: string;
}): Step {
  const step: Step = { type: "assertVisible" };
  step.selector = selectorValue({ text: opts.text, id: opts.id, selector: opts.selector });
  if (opts.timeoutMs != null) step.timeout = opts.timeoutMs;
  if (opts.optional) step.optional = true;
  if (opts.label != null) step.label = opts.label;
  return step;
}

export function assertNotVisible(opts: {
  text?: string;
  id?: string;
  selector?: ElementSelector;
  timeoutMs?: number;
  label?: string;
}): Step {
  const step: Step = { type: "assertNotVisible" };
  step.selector = selectorValue({ text: opts.text, id: opts.id, selector: opts.selector });
  if (opts.timeoutMs != null) step.timeout = opts.timeoutMs;
  if (opts.label != null) step.label = opts.label;
  return step;
}

export function launchApp(
  appId: string,
  opts?: { clearState?: boolean; stopApp?: boolean; label?: string },
): Step {
  const step: Step = { type: "launchApp", appId };
  if (opts?.clearState != null) step.clearState = opts.clearState;
  if (opts?.stopApp != null) step.stopApp = opts.stopApp;
  if (opts?.label != null) step.label = opts.label;
  return step;
}

export function stopApp(appId: string, label?: string): Step {
  const step: Step = { type: "stopApp", appId };
  if (label != null) step.label = label;
  return step;
}

export function clearState(appId: string, label?: string): Step {
  const step: Step = { type: "clearState", appId };
  if (label != null) step.label = label;
  return step;
}

export function openLink(link: string, label?: string): Step {
  const step: Step = { type: "openLink", link };
  if (label != null) step.label = label;
  return step;
}

File: client/python/maestro_runner/commands.py
"""Command builders — produce Go step JSON for the REST API."""

from __future__ import annotations

from typing import Any

from maestro_runner.models import ElementSelector


def _selector_value(
    *,
    text: str | None = None,
    id: str | None = None,
    index: int | None = None,
    selector: ElementSelector | None = None,
    enabled: bool | None = None,
    checked: bool | None = None,
    focused: bool | None = None,
    selected: bool | None = None,
) -> str | dict[str, Any]:
    """Build a selector value (string for text-only, object otherwise)."""
    d: dict[str, Any] = {}
    if selector is not None:
        d.update(selector.to_dict())
    if text is not None:
        d["text"] = text
    if id is not None:
        d["id"] = id
    if index is not None:
        d["index"] = str(index)
    if enabled is not None:
        d["enabled"] = enabled
    if checked is not None:
        d["checked"] = checked
    if focused is not None:
        d["focused"] = focused
    if selected is not None:
        d["selected"] = selected
    # Compact form: text-only selector → plain string
    if list(d.keys()) == ["text"]:
        return str(d["text"])
    return d


def tap_on(
    *,
    text: str | None = None,
    id: str | None = None,
    index: int | None = None,
    selector: ElementSelector | None = None,
    long_press: bool = False,
    wait_until_visible: bool | None = None,
    retry_if_no_change: bool | None = None,
    enabled: bool | None = None,
    checked: bool | None = None,
    focused: bool | None = None,
    selected: bool | None = None,
    optional: bool = False,
    timeout: int | None = None,
    label: str | None = None,
) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "tapOn"}
    step["selector"] = _selector_value(
        text=text, id=id, index=index, selector=selector,
        enabled=enabled, checked=checked, focused=focused, selected=selected,
    )
    if long_press:
        step["longPress"] = True
    if wait_until_visible is not None:
        step["waitUntilVisible"] = wait_until_visible
    if retry_if_no_change is not None:
        step["retryTapIfNoChange"] = retry_if_no_change
    if optional:
        step["optional"] = True
    if timeout is not None:
        step["timeout"] = timeout
    if label is not None:
        step["label"] = label
    return step


def input_text(text: str, *, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "inputText", "text": text}
    if label is not None:
        step["label"] = label
    return step


def erase_text(characters: int | None = None, *, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "eraseText"}
    if characters is not None:
        step["charactersToErase"] = characters
    if label is not None:
        step["label"] = label
    return step


def press_key(code: str, *, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "pressKey", "key": code}
    if label is not None:
        step["label"] = label
    return step


def back(*, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "back"}
    if label is not None:
        step["label"] = label
    return step


def scroll(*, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "scroll"}
    if label is not None:
        step["label"] = label
    return step


def swipe(direction: str, *, duration_ms: int = 400, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {
        "type": "swipe",
        "direction": direction.upper(),
        "duration": duration_ms,
    }
    if label is not None:
        step["label"] = label
    return step


def assert_visible(
    *,
    text: str | None = None,
    id: str | None = None,
    selector: ElementSelector | None = None,
    timeout_ms: int | None = None,
    optional: bool = False,
    label: str | None = None,
) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "assertVisible"}
    step["selector"] = _selector_value(text=text, id=id, selector=selector)
    if timeout_ms is not None:
        step["timeout"] = timeout_ms
    if optional:
        step["optional"] = True
    if label is not None:
        step["label"] = label
    return step


def assert_not_visible(
    *,
    text: str | None = None,
    id: str | None = None,
    selector: ElementSelector | None = None,
    timeout_ms: int | None = None,
    label: str | None = None,
) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "assertNotVisible"}
    step["selector"] = _selector_value(text=text, id=id, selector=selector)
    if timeout_ms is not None:
        step["timeout"] = timeout_ms
    if label is not None:
        step["label"] = label
    return step


def launch_app(
    app_id: str,
    *,
    clear_state: bool | None = None,
    stop_app: bool | None = None,
    label: str | None = None,
) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "launchApp", "appId": app_id}
    if clear_state is not None:
        step["clearState"] = clear_state
    if stop_app is not None:
        step["stopApp"] = stop_app
    if label is not None:
        step["label"] = label
    return step


def stop_app(app_id: str, *, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "stopApp", "appId": app_id}
    if label is not None:
        step["label"] = label
    return step


def clear_state(app_id: str, *, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "clearState", "appId": app_id}
    if label is not None:
        step["label"] = label
    return step


def open_link(link: str, *, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "openLink", "link": link}
    if label is not None:
        step["label"] = label
    return step


def hide_keyboard(*, strategy: str | None = None, label: str | None = None) -> dict[str, Any]:
    step: dict[str, Any] = {"type": "hideKeyboard"}
    if strategy is not None:
        step["strategy"] = strategy
    if label is not None:
        step["label"] = label
    return step


def wait_for_animation_to_end(
    *,
    sleep_ms: int | None = None,
    threshold: float | None = None,
    label: str | None = None,
) -> dict[str, Any]:
    """Build a waitForAnimationToEnd step.

    Args:
        sleep_ms:  Milliseconds to pause between the two comparison screenshots.
                   Longer values catch slow-moving animations.  Defaults to 200 ms
                   on the server side.
        threshold: Maximum pixel-diff percentage (0.0–1.0) still considered static.
                   Lower is stricter.  Defaults to 0.005 (0.5 %) on the server side.
        label:     Optional step label shown in reports.
    """
    step: dict[str, Any] = {"type": "waitForAnimationToEnd"}
    if sleep_ms is not None:
        step["sleepMs"] = sleep_ms
    if threshold is not None:
        step["threshold"] = threshold
    if label is not None:
        step["label"] = label
    return step
